core/trino-main/src/test/java/io/trino/sql/query/TestSessionProperties.java

```diff
@@ -0,0 +1,154 @@
+import static org.junit.jupiter.api.parallel.ExecutionMode.CONCURRENT;
+
+@TestInstance(PER_CLASS)
+@Execution(CONCURRENT)
+final class TestSessionProperties
+{
+    private static final SqlParser SQL_PARSER = new SqlParser();
+    private static final SessionPropertyManager SESSION_PROPERTY_MANAGER = new SessionPropertyManager(
+            ImmutableSet.of(new SystemSessionProperties()),
+            _ -> Map.of("catalog_property", PropertyMetadata.stringProperty("catalog_property", "Test catalog property", "", false)));
+
+    @Test
+    void testParseSystemSessionProperty()
+    {
+        assertThatThrownBy(() -> analyze("WITH SESSION invalid_key = 'invalid_value' SELECT 1"))
+                .hasMessageContaining("line 1:28: Session property invalid_key does not exist");
+
+        assertThatThrownBy(() -> analyze("WITH SESSION query_max_total_memory = 'invalid_value' SELECT 1"))
+                .hasMessageContaining("line 1:39: size is not a valid data size string: invalid_value");
+
+        assertThat(analyze("WITH SESSION query_max_total_memory = '10GB' SELECT 1").getSystemProperties())
+                .isEqualTo(Map.of("query_max_total_memory", "10GB"));
+
+        assertThatThrownBy(() -> analyze("WITH SESSION query_max_total_memory = '10GB', query_max_total_memory = '16GB' SELECT 1"))
+                .hasMessageContaining("line 1:47: Session property query_max_total_memory already set");
+    }
+
+    @Test
+    void testCatalogSessionProperty()
+    {
+        assertThatThrownBy(() -> analyze("WITH SESSION tpch.invalid_key = 'invalid_value' SELECT 1"))
+                .hasMessageContaining("line 1:33: Session property tpch.invalid_key does not exist");
+
+        assertThatThrownBy(() -> analyze("WITH SESSION tpch.catalog_property = true SELECT 1"))
+                .hasMessageContaining("Unable to set session property 'tpch.catalog_property' to 'true': Cannot cast type boolean to varchar");
+
+        assertThat(analyze("WITH SESSION test.catalog_property = 'true' SELECT 1").getCatalogProperties("test"))
+                .isEqualTo(Map.of("catalog_property", "true"));
+
+        assertThat(analyze("WITH SESSION test.catalog_property = CAST(true AS varchar) SELECT 1").getCatalogProperties("test"))
+                .isEqualTo(Map.of("catalog_property", "true"));
+
+        assertThatThrownBy(() -> analyze("WITH SESSION test.catalog_property = 'true', test.catalog_property = 'false' SELECT 1").getCatalogProperties("test"))
+                .hasMessageContaining("line 1:46: Session property test.catalog_property already set");
+    }
+
+    @Test
+    void testInvalidSessionProperty()
+    {
+        assertThatThrownBy(() -> analyze("WITH SESSION test.schema.invalid_key = 'invalid_value' SELECT 1"))
+                .hasMessageContaining("line 1:14: Invalid session property 'test.schema.invalid_key'");
+    }
+
+    private Session analyze(@Language("SQL") String statement)
```

Comment:
Why can't this use a regular QueryRunner? This pattern of assembling part of a query runner makes it harder to maintain and evolve these tests as the underlying classes and interfaces change.

Reply:
It is in the next commit, I've squashed these two together